rasa/server.py

```diff
@@ -1557,33 +1569,58 @@ def _validate_json_training_payload(rjs: Dict) -> None:
             docs=_docs("/api/http-api"),
         )
 
-
 def _training_payload_from_yaml(
     request: Request, temp_dir: Path, file_name: Text = "data.yml"
 ) -> Dict[Text, Any]:
     logger.debug("Extracting YAML training data from request body.")
 
     decoded = request.body.decode(rasa.shared.utils.io.DEFAULT_ENCODING)
     _validate_yaml_training_payload(decoded)
+    request_payload = rasa.shared.utils.io.read_yaml(
+            decoded, reader_type=["safe"]
+        )
+    
+    if "nlu" in request_payload:
+        nlu_path = os.path.join(temp_dir, API_TEMP_DATA_FOLDER, API_TEMP_NLU_YAML_FILE)
+        nlu_data = {'nlu': request_payload["nlu"]}
+        rasa.shared.utils.io.write_yaml(nlu_data, nlu_path)
 
-    training_data = temp_dir / file_name
-    rasa.shared.utils.io.write_text_file(decoded, training_data)
+    if "stories" in request_payload:
+        stories_path = os.path.join(temp_dir, API_TEMP_DATA_FOLDER, API_TEMP_STORIES_YAML_FILE)
+        stories_data = {'stories':request_payload["stories"]}
+        rasa.shared.utils.io.write_yaml(stories_data, stories_path)
+    
+    if "rules" in request_payload:
+        rules_path = os.path.join(temp_dir, API_TEMP_DATA_FOLDER, API_TEMP_RULES_YAML_FILE)
+        rules_data = {'rules':request_payload["rules"]}
+        rasa.shared.utils.io.write_yaml(rules_data, rules_path)
 
+    if "domain" in request_payload:
+        domain_path = os.path.join(temp_dir, API_TEMP_DOMAIN_YAML_FILE)
+        domain_data = request_payload["domain"]
```

Comment:
Not sure with handling all keys to right files and to maintain clean request payload structure, I place certain keys into specific files and form standard training data structure, i.e. 
```yaml
nlu:
  - ...
rules:
  -...
stories:
  -...
config:
  key_config_1
    -...
  key_config_2
    -...
domain:
  key_domain_1
    -...
  key_domain_2
    -...
``` 
rather than

```yaml
nlu:
  - ...
rules:
  -...
stories:
  -...
key_config_1
  -...
key_config_2
  -...
key_domain_1
  -...
key_domain_2
  -...
```